csharp/src/Ice/InputStream.cs

```diff
@@ -1078,11 +1071,43 @@ internal void CheckEndOfBuffer(bool skipTaggedParams)
         }
 
         /// <summary>Reads an encapsulation header from the stream.</summary>
-        /// <returns>The encapsulation header read from the stream.</returns>
-        internal (int Size, Encoding Encoding) ReadEncapsulationHeader()
+        /// <param name="checkFullBuffer">When true, the encapsulation is expected to consume all the bytes of the
+        /// current buffer. When false, bytes can remain in the buffer after the encapsulation.</param>
+        /// <returns>The encapsulation header read from the stream. The size does not include the bytes to the size
+        /// length; it does however include the two byte for the encoding.</returns>
+        internal (int Size, Encoding Encoding) ReadEncapsulationHeader(bool checkFullBuffer)
         {
-            (int size, int sizeLength, Encoding encoding) = _buffer.Span.Slice(Pos).ReadEncapsulationHeader(Encoding);
-            Pos += 2 + sizeLength; // 2 for encoding
+            int size;
+
+            if (OldEncoding)
+            {
+                size = ReadInt();
+                if (size < 0)
```

Comment:
Don't we need at least 4 bytes for the size length here?

```suggestion
                if (size < 4)
```

Reply:
Yes, you're right.